### ISSUE TITLE
num_proc in dataloader affect F1 score in squad

### ISSUE THREAD

**Issue by Yang-YiFan** [huggingface/transformers#19630]
### System Info

transformers 4.18.0 (on CPU)
python 3.9

### Who can help?

@sgugger 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

In the example question answering script, `cd examples/pytorch/question-answering/`

Running the below two cmds will yield different F1 score report (even though their difference is just num workers in dataloader):
1. `python run_qa_no_trainer.py --model_name_or_path csarron/bert-base-uncased-squad-v1 --dataset_name squad --max_seq_length 384 --doc_stride 128 --num_train_epochs 0 --preprocessing_num_workers 1 --output_dir ~/tmp/debug_squad`

2. `python run_qa_no_trainer.py --model_name_or_path csarron/bert-base-uncased-squad-v1 --dataset_name squad --max_seq_length 384 --doc_stride 128 --num_train_epochs 0 --preprocessing_num_workers 4 --output_dir ~/tmp/debug_squad`

### Expected behavior

1. `Evaluation metrics: {'exact_match': 80.90823084200568, 'f1': 88.22754061399627}`
2. `Evaluation metrics: {'exact_match': 76.51844843897824, 'f1': 83.40809222646291}`
They have different results on squad.

**Comment by sgugger** [comment]
Hi there! Could you make your model public? I cannot reproduce this on [distilbert-base-cased-distilled-squad](https://huggingface.co/distilbert-base-cased-distilled-squad) (I get the same scores for the two commands).

**Comment by Yang-YiFan** [comment]
Hi, I believe the one I'm using `csarron/bert-base-uncased-squad-v1` is public. I also tried `bert-large-uncased-whole-word-masking-finetuned-squad` and it's showing the same F1 score mismatch. Thanks!

**Comment by sgugger** [comment]
Ah, just caught the problem in the logs. `preprocessing_num_workers` is the number of workers sent to `Dataset.map`. It should be left as 1 when using a fast tokenizer. When you change it, you change the way the dataset is preprocessed.

To change the number of workers in the dataloader, you should use `dataloader_num_workers`.

**Comment by Yang-YiFan** [comment]
Thanks! The default value is still set to 4 [here](https://github.com/huggingface/transformers/blob/5fda1fbd4625e93d023fe02153ec4a05b26b16cc/examples/pytorch/question-answering/run_qa_no_trainer.py#L111), which should be 1 according to your finding.

**Comment by sgugger** [comment]
Indeed. Do you want to make a PR to fix this?

**Comment by Yang-YiFan** [comment]
Done.
